added azure function for publising and retreiving pilets.

waqasali47 opened the pull request:
# New Pull Request

For more information, see the `CONTRIBUTING` guide.

## Prerequisites

Please make sure you can check the following boxes:

- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project
- [x] All new and existing tests passed

## Type(s) of Changes

### Contribution Type

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue, please reference the issue id)
- [x] New feature (non-breaking change which adds functionality, make sure to open an associated issue first)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes

### Description

[Place a meaningful description here]

### Remarks

[Optionally place any follow-up comments, remarks, observations, or notes here for future reference]

waqasali47 committed:
added azure function for publising and retreiving pilets.

FlorianRappl commented on FeedService/Functions/GetPiletFile.cs:
```diff
@@ -0,0 +1,34 @@
+        {
+            var file = await _piletService.GetPiletFile(fileName, version, name);
+            var response = req.CreateResponse(System.Net.HttpStatusCode.OK);
+            response.Content = new StringContent(file, System.Text.Encoding.UTF8, "text/css");
```
Is this always of type `text/css`? Also if a `.js` file is wanted here? What about binary files (e.g., WASM)? Should we return `StringContent` or actually get the file's content as a `MemoryStream` which is then put as a generic content response?

waqasali47 replied:
What i have test with this is it returns all type of files. i.e. .js or .css. I was able to retrieve the pilet successfully.

FlorianRappl replied:
Yes sure it always *returns*, but the question is if the browser will actually *accept* it. If its the wrong mime type certain assets will be rejected by the browser. Its therefore quite crucial to set the right mime type.

waqasali47 replied:
Shall i  do like this. 
For css i return (text.css)
for JS i return (text/javascript)

FlorianRappl replied:
I'd just detect the mime type via the extension. Potential solution: https://stackoverflow.com/a/14108040/1306682

FlorianRappl commented on FeedService/Functions/Pilet.cs:
```diff
@@ -0,0 +1,51 @@
+                    await _piletService.PublishPilet(fileData);
+                    return new OkObjectResult("Success");
+                }
+                catch (Exception)
```
Right now we only re-throw - should we maybe log something here?

Otherwise I'd propose to just remove the try-catch block.

waqasali47 replied:
Ok. I will remove it for now. Until i find a better logging strategy.

FlorianRappl commented on FeedService/Infrastructure/StorageAccountRepository.cs:
```diff
@@ -0,0 +1,126 @@
+            BlobServiceClient blobServiceClient = new BlobServiceClient(GetConnectionString());
+            BlobContainerClient containerClient = blobServiceClient.GetBlobContainerClient(_piletContainer);
+            var blobs = containerClient.GetBlobs();
+            await foreach (BlobItem item in containerClient.GetBlobsAsync())
```
`await foreach` - I love it :beers:!

FlorianRappl commented on FeedService/Infrastructure/StorageAccountRepository.cs:
```diff
@@ -0,0 +1,126 @@
+                BlobClient bc = containerClient.GetBlobClient(piletName + "/" + version + ".zip");
+                if (await bc.ExistsAsync())
+                {
+                    using (var mem = new MemoryStream())
```
Always getting the full ZIP for each file that is demanded sounds overly complex and slow. Any reason for not uploading *all* files from the archive and just getting the demanded file?

waqasali47 replied:
Right now i am uploading the pilet files as ZIP filder. Because it dosn't make sense to upload single files to storage account as a blob. What i am thinking here to use some kind of intermediate where we keep the files, like Redis cache. Or maybe azure CDN but i am not sure about that yet.

FlorianRappl replied:
> Because it dosn't make sense to upload single files to storage account as a blob.

Can you elaborate why? As far as I know you are not paying by the number of files. But you are paying per compute time on Azure Functions, which increases by having more bytes to download + process (unzip) in this case. Maybe I am wrong here, I'd just love to understand the reasoning.

waqasali47 replied:
I initially thought that it makes sense to put all files in a zip folder, to save space. But you are right that the computation may take longer ( although, i am only opening the zip for a specific file) And it doesn't make sense that each file downloads the whole zip. Maybe i should upload the files as is, in a folder. and then retrieve them one by one. So retrieve only that which is needed.  So i would be something like pilet-name>version>the package folder of pilet. But then I am not sure what happens when we want to retrieve all the pilets.

waqasali47 replied:
What do you say?

FlorianRappl replied:
I'd say upload them already in the right folder structure (in a folder `{pilet-name}/{pilet-version}`). Then you can retrieve only the demanded file via `{pilet-name}/{pilet-version}/{file-path}`.

Regarding the all pilets problem: Well, even if somebody wants to get all pilets all file requests would (I think) still be individual. So this should not matter much. Or am I missing something?

waqasali47 replied:
What i meant was the Get call that this returns https://feed.piral.cloud/api/v1/pilet
It will be hard to return the same format when you call individual files.

FlorianRappl replied:
Hm not sure I follow. Calling `https://feed.piral.cloud/api/v1/pilet` returns the metadata where only a single file is linked, which is the main JS file. So this would lead to `{website-host}/api/{pilet-name}/{pilet-version}/{file-path}`, which we can resolve against the blob storage.

FlorianRappl commented on FeedService/Infrastructure/StorageAccountRepository.cs:
```diff
@@ -0,0 +1,126 @@
+                                        using (var copyMem = new MemoryStream())
+                                        {
+                                            await stream.CopyToAsync(copyMem);
+                                            file = Encoding.Default.GetString(copyMem.ToArray()).Replace("\0", string.Empty); ;
```
While most files will be strings usually files can be anything (and binary).

FlorianRappl commented on FeedService/Models/PackageJson.cs:
```diff
@@ -0,0 +1,52 @@
+        public string Version { get; set; }
+        public bool Preview { get; set; }
+        public string Custom { get; set; }
+        [JsonPropertyName("authoinfo")]
```
I think you meant `author` here.

waqasali47 replied:
Yes. :D Let me update it.

FlorianRappl commented on FeedService/Models/PackageJson.cs:
```diff
@@ -0,0 +1,52 @@
+        public string Custom { get; set; }
+        [JsonPropertyName("authoinfo")]
+        public Author AuthorInfo { get; set; }
+        public string Main { get; set; }
```
I'd say for consistency reasons you should have a `JsonPropertyName("main")` here.

FlorianRappl commented on FeedService/Models/PackageJson.cs:
```diff
@@ -0,0 +1,52 @@
+        [JsonPropertyName("authoinfo")]
+        public Author AuthorInfo { get; set; }
+        public string Main { get; set; }
+        public string Lincense { get; set; }
```
Typo `License` (and I'd add a `JsonPropertyName("license")` here)

FlorianRappl commented on FeedService/Models/PackageJson.cs:
```diff
@@ -0,0 +1,52 @@
+        public Author AuthorInfo { get; set; }
+        public string Main { get; set; }
+        public string Lincense { get; set; }
+        public DevDependencies devDependencies { get; set; }
```
These types (for `*Dependencies`) are wrong. They should `IDictionary<string, string>`. You could make an alias called `using Dependencies = IDictionary<string, string>`.

FlorianRappl commented on FeedService/Services/PiletService.cs:
```diff
@@ -0,0 +1,173 @@
+                var name = ExtractName(pilet);
+                /// This service runs locally. If you want to publish it
+                //you need to handle change this URL accordingly.
+                var url = "http://localhost:7071/api/files/" + name + "/" + version + "/index.js";
```
For the base path we could just use an environment variable.

Usually, this is already set up properly using `Environment.GetEnvironmentVariable("WEBSITE_HOSTNAME")`.

FlorianRappl reviewed (commented):
Awesome - love the direction where this is going!

waqasali47 committed:
removed try catch and update json property for author.

waqasali47 committed:
add environment variable

waqasali47 committed:
added json property.

waqasali47 committed:
removed file zipping logic. Updated functions endpoints.

Updated the function endpoints and Changed to logic to upload files to blob storage. Now we only retrieve files that are needed.

waqasali47 committed:
added connection string comment.

FlorianRappl commented on FeedService/Models/PackageJson.cs:
```diff
@@ -0,0 +1,51 @@
+    public string Version { get; set; }
+    public bool Preview { get; set; }
+    public string Custom { get; set; }
+    [JsonPropertyName("authorinfo")]
```
The name is just `author`

FlorianRappl commented on FeedService/Models/PackageJson.cs:
```diff
@@ -0,0 +1,51 @@
+    [JsonPropertyName("version")]
+    public string Version { get; set; }
+    public bool Preview { get; set; }
+    public string Custom { get; set; }
```
I think for consistency it would make sense to decorate all properties here with the right `JsonPropertyName`. What do you think?

FlorianRappl commented on FeedService/Functions/GetPiletFile.cs:
```diff
@@ -0,0 +1,35 @@
+    {
+      var file = await _piletService.GetPiletFile(fileName, version, name);
+      var provider = new FileExtensionContentTypeProvider();
+      string contentType;
```
Alternatively, you could introduce the variable via `out var contentType` in the next line.

FlorianRappl commented on FeedService/Services/PiletService.cs:
```diff
@@ -0,0 +1,130 @@
+      {
+        var version = ExtractVersion(pilet);
+        var name = ExtractName(pilet);
+        var url = "http://" + Environment.GetEnvironmentVariable("WEBSITE_HOSTNAME") + "/api/pilet/" + name + "/" + version + "/index.js";
```
Hm, I think we can also get the right scheme, right? Because usually Azure Functions should be `https` (except using the local emulator).

FlorianRappl commented on FeedService/Services/PiletService.cs:
```diff
@@ -0,0 +1,130 @@
+
+    private string ExtractVersion(string pilet)
+    {
+      return pilet.Split('/')[1].Substring(0, 5);
```
How is the `0, 5` derived here? Is this due to `0.1.2`? But what if the version is `0.0.1001` or `12.19.201`?

FlorianRappl reviewed (commented):
Awesome - looks really great!

Just a couple of questions and suggestions from my side.

waqasali47 committed:
added env specific call for https  and some refactoring.

FlorianRappl reviewed (approved):
LGTM - awesome work :beers:!
